src/crimson/osd/replicated_backend.cc

```diff
@@ -38,31 +40,337 @@ ReplicatedBackend::_read(const hobject_t& hoid,
   return store->read(coll, ghobject_t{hoid}, off, len, flags);
 }
 
+
+/*
+ * High level transaction.
+ *
+ * Due to how clones, renames (and EC) are handled in RADOS, crimson
+ * needs something similar to PGTransaction. Unlike it, HLTransaction
+ * is just an extension (also from the memory arrangement's POV) to
+ * the low-level os::Transaction. These jettisonable memory areas
+ * helps to reorder operations.
+ */
+namespace crimson::osd {
+class HLTransaction {
+public:
+  using Op = ceph::os::Transaction::Op;
+
+  HLTransaction(ceph::os::Transaction&& txn)
+    : txn(txn) {
+    if (!this->txn.empty()) {
+      // doing this in the transition period. The goal is to have
+      // full os:Transaction-like interface and make the data loc
+      // info when e.g. `clone()` or `write()` are called.
+      scan_data_locators();
+    }
+  }
+
+  auto* begin() {
+    return reinterpret_cast<Op*>(txn.op_bl.c_str());
+  }
+
+  auto* end() {
+    return begin() + txn.get_num_ops();
+  }
+
+  decltype(auto) get_object_index() const {
+    return txn.get_object_index();
+  }
+
+  void reorder_back(Op& op) {
+    const auto idx = std::distance(begin(), &op);
+    std::swap(*std::next(std::begin(*this), idx),
+              *std::next(std::begin(*this), ordered_tail_idx));
+    std::swap(op_data_locators[idx], op_data_locators[ordered_tail_idx]);
+    ++ordered_tail_idx;
+  }
+
+  ceph::os::Transaction& get_lltransaction() && {
```

Comment:
get_futurized_store_transaction() -- again, let's just name things based on the interface that consumes them.